hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java

```diff
@@ -607,40 +621,77 @@ protected void handleFlush(boolean close) throws IOException {
+      if (totalWriteChunkLength < writtenDataLength) {
+        Preconditions.checkArgument(currentBuffer.position() > 0);
+
+        // This can be a partially filled chunk. Since we are flushing the buffer
+        // here, we just limit this buffer to the current position. So that next
+        // write will happen in new buffer
+        updateWriteChunkLength();
+        updatePutBlockLength();
+        if (currentBuffer.hasRemaining()) {
+          if (allowPutBlockPiggybacking) {
+            putBlockResultFuture = writeChunkAndPutBlock(currentBuffer, close);
+          } else {
+            writeChunk(currentBuffer);
+            putBlockResultFuture = executePutBlock(close, false);
+          }
+          if (!close) {
+            // reset current buffer.
+            currentBuffer = null;
+            currentBufferRemaining = 0;
+          }
         } else {
-          writeChunk(currentBuffer);
-          updateFlushLength();
-          executePutBlock(close, false);
+          putBlockResultFuture = executePutBlock(close, false);
+          // set lastFuture.
         }
+      } else if (totalPutBlockLength < totalWriteChunkLength) {
+        // There're no pending written data, but there're uncommitted data.
+        updatePutBlockLength();
+        putBlockResultFuture = executePutBlock(close, false);
+      } else if (close) {
+        // forcing an "empty" putBlock if stream is being closed without new
+        // data since latest flush - we need to send the "EOF" flag
+        updatePutBlockLength();
+        putBlockResultFuture = executePutBlock(true, true);
       } else {
-        updateFlushLength();
-        executePutBlock(close, false);
+        LOG.debug("Flushing without data");
+      }
+
+      if (putBlockResultFuture != null) {
+        this.lastFlushFuture = watchForCommitAsync(putBlockResultFuture);
+      }
+      toWaitFor = this.lastFlushFuture;
+    } // End of synchronized block.
+
+    if (toWaitFor != null) {
+      LOG.debug("Waiting for flush");
+      try {
+        toWaitFor.get();
+      } catch (ExecutionException ex) {
+        if (ex.getCause() instanceof FlushRuntimeException) {
+          throw ((FlushRuntimeException) ex.getCause()).cause;
+        } else {
+          throw ex;
+        }
       }
-    } else if (close) {
-      // forcing an "empty" putBlock if stream is being closed without new
-      // data since latest flush - we need to send the "EOF" flag
-      executePutBlock(true, true);
     }
-    waitOnFlushFutures();
-    watchForCommit(false);
-    // just check again if the exception is hit while waiting for the
-    // futures to ensure flush has indeed succeeded
+    LOG.debug("Flush done.");
+  }
 
-    // irrespective of whether the commitIndex2flushedDataMap is empty
-    // or not, ensure there is no exception set
-    checkOpen();
+  private CompletableFuture<Void> watchForCommitAsync(CompletableFuture<PutBlockResult> putBlockResultFuture) {
```

Comment:
`watchForCommit` is synced.  The code below requires a thread to wait for the `replyFuture.get()` in `XceiverClientRatis.watchForCommit(index)`.  We should check which thread it will use (print the thread name) and if it is okay for that thread to wait.

A better way to to return the `replyFuture` in the first place but it will need more refactoring.  We may do it separately.

Reply:
Filed HDDS-11174.

Reply:
The waiting thread is named `client-write-TID-{}`. Which is from `responseExecutor` in `BlockOutputStream`. It take over handling the future return from ratis client via `thenApplyAsync`.

```
  CompletableFuture<PutBlockResult> executePutBlock(boolean close,
      asyncReply = putBlockAsync(xceiverClient, blockData, close, tokenString);
      CompletableFuture<ContainerCommandResponseProto> future = asyncReply.getResponse();
      flushFuture = future.thenApplyAsync(e -> {
        try {
          validateResponse(e);
        } catch (IOException sce) {
          throw new CompletionException(sce);
        }
        // if the ioException is not set, putBlock is successful
        if (getIoException() == null && !force) {
          handleSuccessfulPutBlock(e.getPutBlock().getCommittedBlockLength(),
              asyncReply, flushPos, byteBufferList);
        }
        return e;
      }, responseExecutor)
```
The thread we use to wait for `putBlock` result and then call `watchForCommit` is from the designated for handling putBlock response. It's not the thread from ratis-client internal. That should be fine.